### PR TITLE
Insert a breakpoint instruction after throwing calls

### PR DESCRIPTION
**Pull request by VSadov**
A follow up for https://github.com/dotnet/runtime/pull/100900

See discussion at: https://github.com/dotnet/runtime/pull/100900#discussion_r1562851304

Instead of inserting a breakpoint at the end of `BBJ_THROW` blocks that end on a throwing call, we insert a breakpoint after throwing calls.

The change should end up with roughly the same number of added breakpoint instructions, since a BBJ_THROW nearly always ends with a throwing call anyways. However, this approach would be more robust in rare situations when a throwing call is followed by some extra code, which may potentially be removed at later phases.

### CODE REVIEW

**Comment by VSadov** [comment]
A fairly trivial change - mostly moving insertion of BRK from one place to another and covering a few places that older approach missed.

However, somehow it causes a crossgen failure when building clr subset:

```
  Opening /home/vsadov/dotnet002/runtime/artifacts/mibc/linux/arm/data/DotNet_Adhoc.mibc
  Opening /home/vsadov/dotnet002/runtime/artifacts/mibc/linux/arm/data/DotNet_FirstTimeXP.mibc
  Opening /home/vsadov/dotnet002/runtime/artifacts/mibc/linux/arm/data/DotNet_FSharp.mibc
  Opening /home/vsadov/dotnet002/runtime/artifacts/mibc/linux/arm/data/DotNet_HelloWorld.mibc
  Opening /home/vsadov/dotnet002/runtime/artifacts/mibc/linux/arm/data/DotNet_OrchardCore.mibc
  Opening /home/vsadov/dotnet002/runtime/artifacts/mibc/linux/arm/data/DotNet_TechEmpower.mibc
  Generated /home/vsadov/dotnet002/runtime/artifacts/bin/coreclr/linux.arm.Checked/StandardOptimizationData.mibc
  Validated /home/vsadov/dotnet002/runtime/artifacts/bin/coreclr/linux.arm.Checked/StandardOptimizationData.mibc
  Generating native image of System.Private.CoreLib for linux.arm.Checked. Logging to
  /home/vsadov/dotnet002/runtime/dotnet.sh /home/vsadov/dotnet002/runtime/artifacts/bin/coreclr/linux.arm.Checked/arm64/crossgen2/tools/crossgen2.dll -o:/home/vsadov/dotnet002/runtime/artifacts/bin/coreclr/linux.arm.Checked/System.Private.CoreLib.dll -r:/home/vsadov/dotnet002/runtime/artifacts/bin/coreclr/linux.arm.Checked/IL/*.dll --targetarch:arm --targetos:linux -m:/home/vsadov/dotnet002/runtime/artifacts/bin/coreclr/linux.arm.Checked/StandardOptimizationData.mibc --embed-pgo-data -O --verify-type-and-field-layout /home/vsadov/dotnet002/runtime/artifacts/bin/coreclr/linux.arm.Checked/IL/System.Private.CoreLib.dll --perfmap-format-version:1 --perfmap --perfmap-path:/home/vsadov/dotnet002/runtime/artifacts/bin/coreclr/linux.arm.Checked/
  /home/vsadov/dotnet002/runtime/src/coreclr/jit/emit.h:2406
  Assertion failed 'offset < (emitTotalHotCodeSize + emitTotalColdCodeSize)' in 'System.Threading.EventWaitHandle:TryOpenExisting(System.String,byref):ubyte' during 'Emit code' (IL size 11; hash 0xae6c977d; FullOpts)

  /home/vsadov/dotnet002/runtime/src/coreclr/jit/emit.h:2406
  Assertion failed 'offset < (emitTotalHotCodeSize + emitTotalColdCodeSize)' in 'System.Threading.Semaphore:TryOpenExisting(System.String,byref):ubyte' during 'Emit code' (IL size 11; hash 0x32a77fca; FullOpts)
```

If crossgen is disabled, product builds and tests run. 
I suspect, perhaps it is something with applying `.mibs`

**Comment by VSadov** [comment]
The failures are indeed around places where we could pick a few cases not covered before. 

`EventWaitHandle.TryOpenExisting`

```cs
        [SupportedOSPlatform("windows")]
        public static bool TryOpenExisting(string name, [NotNullWhen(true)] out EventWaitHandle? result) =>
            OpenExistingWorker(name, out result!) == OpenExistingResult.Success;
```
and `OpenExistingWorker` throws PNSE on Unix instead of returning a value
```cs
        private static OpenExistingResult OpenExistingWorker(string name, out EventWaitHandle? result)
        {
            throw new PlatformNotSupportedException(SR.PlatformNotSupported_NamedSynchronizationPrimitives);
        }
```

Since the throwing call is an argument to a comparison, it is very possible that it had some dead code after the call and was not seen by the previous approach, since the call is not at the end of the block.  

But weird thing is that this does not impact normal execution of tests and sems to only impact crossgen. Crossgen has extra inputs that are not specific to the current code, so I think maybe this is some disconnect between expected/actual.

File: eng/Subsets.props [hunk 1]
```diff
@@ -394,7 +394,7 @@
       Test="true" Category="tools"/>
   </ItemGroup>
 
-  <ItemGroup Condition="$(_subset.Contains('+clr.nativecorelib+'))">
+  <ItemGroup Condition="$(_subset.Contains('+clr.nativecorelib+')) and '$(TargetArchitecture)' != 'arm'">
```
Comment (VSadov) [inline eng/Subsets.props:397]: A temporary workaround for the crossgen failure. Maybe there is a better solution...

**Comment by AndyAyersMS** [comment]
Did you look into breaking trees with no return calls and discarding anything that comes later? Would be nice if there weren't special cases like this.

**Comment by VSadov** [comment]
> Did you look into breaking trees with no return calls and discarding anything that comes later? Would be nice if there weren't special cases like this.

From the discussion https://github.com/dotnet/runtime/pull/100900#discussion_r1562851304 it seemed that there are cases where this cannot be easily fixed and adding a breakpoint after calls is a more robust strategy.

**Comment by AndyAyersMS** [comment]
> I have seen code that tries to remove everything in a block after a throwing call, but it looks like it does not always work. Perhaps some cases materialize too late.

Morph does this, but at statement level granularity, so if the throwing call is a subtree it may leave stuff up above sometimes. It may be painful to clean up in tree form. Once we get to linear form it should be simpler to get rid of whatever comes after the call.

**Comment by VSadov** [comment]
> Once we get to linear form it should be simpler to get rid of whatever comes after the call.

That sounds like just stop emitting the block after seeing a no-return call. Is that the idea?

**Comment by AndyAyersMS** [comment]
> > Once we get to linear form it should be simpler to get rid of whatever comes after the call.
> 
> That sounds like just stop emitting the block after seeing a no-return call. Is that the idea?

More or less, yes. 

There may be a bit more to it than that, if the throwing call's use has other uses you may need to mark them as unused or something.

**Comment by VSadov** [comment]
Would be an interesting thing to try. 

But we will still insert BRK in `CodeGen::genCall`, right?

One reason for this could be that `CodeGen::genCall` may try to emit more instructions after the call - seems like it may try to swap return registers between ABI and expected, there is something SIMD related, etc... Not sure how much of that could be possibly applicable to a no-return call, but seems like some of that might.

Even if we insert BRK in codegen, `CodeGen::genCall` would need to know to ensure the call instruction is the last one if the call is no-return. Then why not just insert break point there and be sure that nothing can sneak in between.

**Comment by VSadov** [comment]
> Then why not just insert break point there and be sure that nothing can sneak in between.

Actually that is not very important. The extra instructions could be redundant, but if the breakpoint comes after it is still ok.

**Comment by VSadov** [comment]
Interesting failure. We morph binary operators that are statically overflowing into `comma(throw(OVF), 0)` .

https://github.com/dotnet/runtime/blob/b35b67d24c211a73a1eed7213aef245dd0e0cbcc/src/coreclr/jit/gentree.cpp#L16327

It looks like 0 literals can consume temps, so if I just drop the rest of the block, I get 
```
Assertion failed 'tmpGetCount == 0' in 'ILGEN_0xc2f3d0c8:Method_0x5cc2(double):int'
```
Is there a way to manufacture a value that would not use any kind of temp? 
Assuming that we know that we will not emit it.

Alternatively - maybe we can tell that this instance of CORINFO_HELP_OVERFLOW actually returns nonvoid type and use it as-is, not wrap it in a comma.  
(I tried, but got some asserts about unmorphed nodes, does it have to be comma?)

Also the problem with temps could be more general than just this case. I am not sure yet.

**Comment by VSadov** [comment]
I think this is obsolete now. We do the same as a result of interpretability change which inserts breakpoints in these locations.

**Pull request closed without merging**